feat: Docker-compose: add docker compose with App+PostgreSQL

mcastellin opened the pull request:
Creating docker-compose file to spin up a postgresql service with the application.

- Application can build from docker-compose
- Configured Postgresql DB with persistent volume
- DB password can be provided with environment file

Fix #72

XaviTorello commented on docker-compose.yml:
```diff
@@ -0,0 +1,35 @@
+      - db
+    ports:
+      - 8000:8000
+
```
Current code should be mounted into the container to simplify and enhance development workflow.

It can be achieved with something like:
```
    volumes:
      - .:/anem-per-feina/
```

With this, the base image is not needed to be recreated for each change. Also Djangos reload-on-changes default behavior will be preserved.

fullonic replied:
Would be great to the app volume mounted as @XaviTorello suggest,  so we can development already in a docker environment.

XaviTorello commented on docker-compose.yml:
```diff
@@ -0,0 +1,35 @@
+    ports:
+      - 8000:8000
+
+    environment:
```
Composition envs should be also loaded here to ensure that same values are passed to both services

```
    env_file:
      - ./docker/secrets.env
```

XaviTorello commented on docker/secrets.env.sample:
```diff
@@ -0,0 +1,4 @@
+# Use this file to store secrets used in Docker containers
```
Maybe would be clear if we just define an `docker/.env` file to also store or override default vars.

The idea is to load it in our composition to patch "realtime" this values in all related services.

XaviTorello replied:
I mean, I would also like to see here any composition vars, like `POSTGRES_HOST=db`, `POSTGRES_USER`, ...

What do you think @mcastellin?

mcastellin replied:
@XaviTorello you can now provide thos variables from the `docker/.env` file

XaviTorello commented on docker-compose.yml:
```diff
@@ -0,0 +1,35 @@
+      #   and generate your container secrets
+      - ./docker/secrets.env
+    environment:
+      POSTGRES_DB: anemperfeina
```
Maybe will be cleaner if we isolate it inside our docker `.env` file

XaviTorello commented on Dockerfile:
```diff
@@ -1,14 +1,31 @@
+# drop root privileges when running application in container
+USER ${USR}
+
+RUN (cat ${DOTENV}; echo; cat ./docker/secrets.env) > .env
```
I would not patch .env definition into the image (as a constant), maybe it's better to define it "live" with `env_file` in our composition.

This image should be used also to deploy our service to k8s, CI, ...

XaviTorello reviewed (changes requested):
Many thanks for your effort @mcastellin, I just would to share some ideas with you.

Basically I would love to store all the env vars that we'll use to patch the Docker environment in one same place.

Also, I think that is a good idea to mount current code status in our composition to provide an improved development experience without the need to recreate the image for our base container.

I totally loved the integration of `wait-for-it` :)

Thoughts?

fullonic commented on .dockerignore:
```diff
@@ -0,0 +1,11 @@
+*.pyo
+*.pyd
+.Python
+env
```
could we add as well `venv`? Is a very common way to name the virtual environment folder.

jbagot replied:
yes, agree, `venv` should be ignored as well

fullonic replied:
@mcastellin Just now I'm seeing that only `db.sqlite3` is being ignored. However the development version of `sqlite` was renamed to `anem-per-feina.sqlite`.

https://github.com/GeeksCAT/anem-per-feina/blob/379c47df481f2aaf8813543dae2011a594777857/.env.dev.sample#L4

mcastellin replied:
@fullonic I have updated the ignore statement to include any file ending by `.sqlite3` or `.sqlite`

fullonic replied:
Great job @mcastellin :)

jbagot commented on docker/wait-for-it.sh:
```diff
@@ -0,0 +1,182 @@
+#!/usr/bin/env bash
```
It's fine this. But I normally use `apt install wait-for-it` in the Dockerfile and then you can use it in the docker-compose and you don't need to have this script in the repo.

mcastellin commented:
@XaviTorello @jbagot I have addressed all points to make the docker-compose usable for the development process.
I also modified the dockerfile to use the `slim-buster` version as discussed in #60 if we merge this one, I think the other can be closed since the change is already included here.

I have pushed the code changes and will now also provide some documentation of what you need to do in order to use the image for development.

mcastellin commented on .env.dev.sample:
```diff
@@ -1,3 +1,3 @@
-; To use dev env ini file.
-; cp .env.dev.sample .env
+# To use dev env ini file.
```
I have replaced with `#` since the the semicolon was showing warning messages when running the application

mcastellin commented on Dockerfile:
```diff
@@ -1,14 +1,36 @@
+# drop root privileges when running application in container
+USER ${USR}
+
+# the Django environment file.
+# if this value is specified the .env file will be replaced at build time
+ARG DOTENV=
+# configured .env file at build time
+RUN ./docker/config-env.sh
```
By default the .env file will not be created or overridden although I have made it possible by using the `DOTENV` build argument in case we want to build the image for staging or production with custom env files.

jbagot replied:
For now I'm ok with it. Maybe tomorrow I'll do some modifications on top of it. Great work, thanks.

jbagot commented on README.md:
```diff
@@ -53,6 +53,56 @@ Run server:
+
+2. Add `docker/.env` file for env override when running in container
+
+   `cp docker/.env.dev.sample dvocker/.env`
```
typo docker

jbagot commented:
When you try to run the docker-compose doesn't work because the development settings tries to import `.base`. I made a PR solving this bug. Once my PR is merged this docker-compose will work. @mcastellin @XaviTorello

jbagot commented:
When the typo in the readme will be fixed for me will be approved. After that with this basis I'll do some changes on top of it.

mcastellin commented:
Typo corrected, awaiting final review.

@jbagot Thanks for addressing the issue with the `.base` import. I didn't notice that one! 
What is it about the `config-env.sh` that doesn't convince you?

jbagot commented:
> Typo corrected, awaiting final review.
> 
> @jbagot Thanks for addressing the issue with the `.base` import. I didn't notice that one!
> What is it about the `config-env.sh` that doesn't convince you?

The `config-env.sh` and all the settings approach I think can be easier, I want to merge this PR and then work on that. For example, I would use only one .env, and this .env file has to be there, if the .env doesn't exist, fails. And don't override settings and anything. As simple as read the .env file in the docker-compose. Then in the k8s we will have to think how to manage the secrets and envs. I have couple of ideas but we can discuss it later.
But as I said I'm completely fine with this PR. The only thing is that now @XaviTorello created the pipeline and black is complaining in the `settings/development.py` file. Could you fix it following black please? And then I'll merge it. @mcastellin

jbagot commented on jobs/settings/development.py:
```diff
@@ -1,5 +1,16 @@
 from . import *  # noqa
 
+DATABASES = {
+    'default': {
```
Black is complaining because have to be double quote. Run black and it will change everything as it wants.

mcastellin commented:
`development.py` is now formatted. Thanks @jbagot for the explaination
